File: 강의 자료/02-알고리즘/03~04-정렬/문제별 코드/15970-화살표 그리기/solution.py
import sys
n = int(sys.stdin.readline())

a = [[] for _ in range(n + 1)]

for i in range(n):
    coord, color = map(int, sys.stdin.readline().split())
    a[color].append(coord)

def toLeft(color, i):
    if i == 0:
        return 1000000
    return a[color][i] - a[color][i - 1]

def toRight(color, i):
    if i + 1 == len(a[color]):
        return 1000000
    return a[color][i + 1] - a[color][i]


ans = 0
for color in range(1, n + 1):
    a[color].sort()
    for i in range(len(a[color])):
        ans += min(toLeft(color, i), toRight(color, i))

print(ans)

File: 강의 자료/02-알고리즘/03~04-정렬/문제별 코드/15970-화살표 그리기/solution.cpp
#include <iostream>
#include <algorithm>
#include <vector>
typedef long long int ll;
using namespace std;

#define NM 5005

int N;
vector<int> a[NM];

void input() {
	cin >> N;
	for (int i = 1; i <= N; i++) {
		int coord, color;
		cin >> coord >> color;
		a[color].push_back(coord);
	}
}

int toLeft(int color, int x) {
	if (x == 0) return 0x7fffffff;
	return a[color][x] - a[color][x - 1];
}

int toRight(int color, int x) {
	if (x + 1 == a[color].size()) return 0x7fffffff;
	return a[color][x + 1] - a[color][x];
}

void pro() {
	int ans = 0;
	for (int color = 1; color <= N; color++) {
		sort(a[color].begin(), a[color].end());
		for (int i = 0; i < a[color].size(); i++) {
			int left = toLeft(color, i);
			int right = toRight(color, i);
			ans += min(left, right);
		}
	}
	cout << ans;
}


int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	pro();
	return 0;
}